File: code/python/estimate_demand.py
"""
Nested logit demand estimation — Python port of Julia estimate_demand_v3.jl.

Two-stage:
  Stage 1: BFGS with BHHH initialization (β only, λ=1 fixed)
  Stage 2: L-BFGS-B joint optimization (β + λ)

Key design:
  V_0 = β'X_0 (NOT normalized to 0)
  Weights normalized to mean 1 globally
  Dict-based HH grouping (no sorting)
  Single-threaded BLAS for determinism
"""

import os
os.environ["OPENBLAS_NUM_THREADS"] = "1"
os.environ["MKL_NUM_THREADS"] = "1"

import glob
import csv
import time
import numpy as np
from scipy.optimize import minimize
from datetime import datetime

# =========================================================================
# Covariates (must match Julia BASE_COVARS + structural covars)
# =========================================================================

COVARS = [
    "premium", "penalty_own", "premium_sq",
    "silver", "bronze", "hh_size_prem",
    "any_0to17_prem", "FPL_250to400_prem", "FPL_400plus_prem",
    "any_black_prem", "any_hispanic_prem", "hmo", "hsa",
    "Anthem", "Blue_Shield", "Kaiser", "Health_Net",
    "Anthem_silver", "BS_silver", "Kaiser_silver", "HN_silver",
    "Anthem_bronze", "BS_bronze", "Kaiser_bronze", "HN_bronze",
    "assisted_silver", "assisted_bronze", "assisted_gold", "assisted_plat",
    "commission_broker", "v_hat_commission",
]
K = len(COVARS)


# =========================================================================
# Data structures
# =========================================================================

class HHGroup:
    __slots__ = ["ins_rows", "unins_row", "chosen_row", "chose_insured", "weight"]
    def __init__(self, ins_rows, unins_row, chosen_row, chose_insured, weight):
        self.ins_rows = ins_rows        # numpy int array of insured row indices
        self.unins_row = unins_row      # int
        self.chosen_row = chosen_row    # int
        self.chose_insured = chose_insured  # bool
        self.weight = weight            # float


class CellData:
    __slots__ = ["X", "groups", "n_hh"]
    def __init__(self, X, groups, n_hh):
        self.X = X              # (n_rows, K) float64
        self.groups = groups    # list of HHGroup
        self.n_hh = n_hh        # int


# =========================================================================
# Build cell data from CSV (dict-based grouping, no sort)
# =========================================================================

def build_cell_data(filepath):
    with open(filepath, "r") as f:
        reader = csv.DictReader(f)
        rows = list(reader)

    n = len(rows)
    if n == 0:
        return None

    # Build X matrix
    X = np.zeros((n, K), dtype=np.float64)
    for k, var in enumerate(COVARS):
        for i in range(n):
            val = rows[i].get(var, "")
            try:
                X[i, k] = float(val) if val != "" else 0.0
            except (ValueError, TypeError):
                X[i, k] = 0.0

    # Group rows by household using dict
    hh_rows_dict = {}
    for i in range(n):
        hh = int(float(rows[i]["household_number"]))
        if hh not in hh_rows_dict:
            hh_rows_dict[hh] = []
        hh_rows_dict[hh].append(i)

    groups = []
    for hh, row_indices in hh_rows_dict.items():
        ins_rows = []
        unins_row = -1
        chosen_row = -1

        for r in row_indices:
            if rows[r]["plan_name"] == "Uninsured":
                unins_row = r
            else:
                ins_rows.append(r)
            if int(float(rows[r]["choice"])) == 1:
                chosen_row = r

        if len(ins_rows) == 0 or unins_row == -1 or chosen_row == -1:
            continue

        chose_insured = rows[chosen_row]["plan_name"] != "Uninsured"
        weight = float(rows[row_indices[0]].get("ipweight", "1"))

        groups.append(HHGroup(
            ins_rows=np.array(ins_rows, dtype=np.int64),
            unins_row=unins_row,
            chosen_row=chosen_row,
            chose_insured=chose_insured,
            weight=weight,
        ))

    if len(groups) == 0:
        return None

    return CellData(X=X, groups=groups, n_hh=len(groups))


# =========================================================================
# Load all cells
# =========================================================================

def load_cells(cell_dir):
    pattern = os.path.join(cell_dir, "cell_*_data.csv")
    csv_files = sorted(glob.glob(pattern))
    print(f"  Found {len(csv_files)} cell files")

    cells = []
    total_hh = 0
    for i, f in enumerate(csv_files):
        cell = build_cell_data(f)
        if cell is not None:
            cells.append(cell)
            total_hh += cell.n_hh
        if (i + 1) % 20 == 0:
            print(f"    Loaded {i + 1} / {len(csv_files)}")

    print(f"  Loaded {len(cells)} cells, {total_hh} HH")
    return cells, total_hh


# =========================================================================
# Normalize weights globally to mean 1
# =========================================================================

def normalize_weights(cells):
    total_w = 0.0
    total_n = 0
    for cell in cells:
        for g in cell.groups:
            total_w += g.weight
            total_n += 1
    global_mean = total_w / total_n
    print(f"  Weights normalized: global mean was {global_mean:.4f}")

    for cell in cells:
        for g in cell.groups:
            g.weight /= global_mean

    return global_mean


# =========================================================================
# NLL + gradient for one cell (per-HH loop — matches Julia exactly)
# =========================================================================

def cell_negll_grad(beta, lam, cell, compute_gradi=False):
    X = cell.X
    n_hh = cell.n_hh

    # V = X @ β for all rows
    V = X @ beta

    negll = 0.0
    grad = np.zeros(K + 1)
    gradi = np.zeros((n_hh, K + 1)) if compute_gradi else None

    for h, g in enumerate(cell.groups):
        ins = g.ins_rows
        w = g.weight

        # Insured nest: logsumexp of V_ins / λ
        V_ins = V[ins]
        V_scaled = np.clip(V_ins / lam, -500, 500)
        max_vs = np.max(V_scaled)
        exp_vs = np.exp(V_scaled - max_vs)
        D = np.sum(exp_vs)
        if D < 1e-300:
            D = 1e-300
        I_val = max_vs + np.log(D)

        # Within-nest shares and weighted averages
        s_jg = exp_vs / D
        X_ins = X[ins]
        x_bar = s_jg @ X_ins          # (K,)
        V_bar = np.dot(s_jg, V_ins)   # scalar

        # V_0 = β'X_0 (NOT zero)
        V_0 = V[g.unins_row]
        X_0 = X[g.unins_row]

        # log denominator: log(exp(λI) + exp(V_0))
        lI = lam * I_val
        mx_d = max(lI, V_0)
        log_denom = mx_d + np.log(np.exp(np.clip(lI - mx_d, -500, 500))
                                   + np.exp(np.clip(V_0 - mx_d, -500, 500)))

        # P(insured)
        P_ins = np.exp(lI - log_denom)
        P_ins = np.clip(P_ins, 1e-15, 1 - 1e-15)

        # Log-likelihood
        if g.chose_insured:
            V_ch = V[g.chosen_row]
            ll_h = V_ch / lam + (lam - 1.0) * I_val - log_denom
        else:
            ll_h = V_0 - log_denom
        negll -= w * ll_h

        # Gradient w.r.t. β
        if g.chose_insured:
            X_ch = X[g.chosen_row]
            g_beta = (X_ch - x_bar) / lam + (1.0 - P_ins) * (x_bar - X_0)
            g_lam = (-V_ch / lam**2 + I_val
                     - (lam - 1.0) * V_bar / lam**2
                     - P_ins * (I_val - V_bar / lam))
        else:
            g_beta = -P_ins * (x_bar - X_0)
            g_lam = -P_ins * (I_val - V_bar / lam)

        grad[:K] -= w * g_beta
        grad[K] -= w * g_lam

        if compute_gradi:
            gradi[h, :K] = -w * g_beta
            gradi[h, K] = -w * g_lam

    # Guard against NaN
    if not np.isfinite(negll):
        negll = 1e20
    grad[~np.isfinite(grad)] = 0.0

    return negll, grad, gradi


# =========================================================================
# Accumulate NLL + gradient across cells
# =========================================================================

def accumulate(theta, cells, compute_gradi=False):
    beta = theta[:K]
    lam = theta[K]

    total_negll = 0.0
    total_grad = np.zeros(K + 1)
    all_gradi = [] if compute_gradi else None

    for cell in cells:
        nll, grad, gradi = cell_negll_grad(beta, lam, cell, compute_gradi)
        total_negll += nll
        total_grad += grad
        if compute_gradi:
            all_gradi.append(gradi)

    return total_negll, total_grad, all_gradi


# =========================================================================
# Stage 1: BFGS-BHHH on β with λ fixed (matches Julia optimize_beta)
# =========================================================================

def optimize_beta(cells, beta_init=None, lam_fixed=1.0,
                  max_iter=2000, tol=1e-6, stptol=1e-10):

    beta = np.zeros(K) if beta_init is None else beta_init.copy()
    theta = np.append(beta, lam_fixed)

    # Initial evaluation with BHHH
    print("  Computing BHHH initialization...")
    negll, full_grad, all_gradi = accumulate(theta, cells, compute_gradi=True)
    grad_beta = full_grad[:K]

    # BHHH matrix (β dimensions only)
    bhhh = np.zeros((K, K))
    for gradi in all_gradi:
        gi_beta = gradi[:, :K]
        bhhh += gi_beta.T @ gi_beta
    del all_gradi

    try:
        Hm1 = np.linalg.inv(bhhh)
    except np.linalg.LinAlgError:
        print("  WARNING: BHHH singular, using scaled identity")
        Hm1 = np.eye(K) * 1e-4

    print(f"  Init: negLL = {negll:.2f}  beta_1 = {beta[0]:.6f}")

    for it in range(1, max_iter + 1):
        old_negll = negll
        old_grad = grad_beta.copy()

        d = -Hm1 @ grad_beta

        # Halving line search
        step = 2.0
        while True:
            step /= 2.0
            if step < stptol:
                break
            theta_try = np.append(beta + step * d, lam_fixed)
            nll_try, _, _ = accumulate(theta_try, cells)
            if nll_try <= old_negll:
                break

        if step < stptol:
            print(f"    beta iter {it:3d}: step too small")
            break

        # Accept
        beta = beta + step * d
        theta = np.append(beta, lam_fixed)
        negll, full_grad, _ = accumulate(theta, cells)
        grad_beta = full_grad[:K]

        # BFGS inverse Hessian update
        incr = step * d
        y = grad_beta - old_grad
        sy = np.dot(incr, y)
        if abs(sy) > 1e-20:
            yHy = y @ Hm1 @ y
            Hm1 = (Hm1
                    + (sy + yHy) / sy**2 * np.outer(incr, incr)
                    - (Hm1 @ np.outer(y, incr) + np.outer(incr, y) @ Hm1) / sy)

        if it <= 3 or it % 10 == 0:
            print(f"    beta iter {it:3d}: negLL = {negll:.2f}  step = {step:.4f}  beta_1 = {beta[0]:.6f}")

        if abs(negll - old_negll) < tol:
            print(f"    beta converged at iter {it}")
            break

    return beta, negll


# =========================================================================
# Optimize λ with β fixed (golden section search, matches Julia)
# =========================================================================

def optimize_lambda(cells, beta, lam_lo=0.01, lam_hi=1.0, tol=1e-4):

    phi = (np.sqrt(5.0) - 1.0) / 2.0  # golden ratio

    a, b = lam_lo, lam_hi
    c = b - phi * (b - a)
    d = a + phi * (b - a)

    theta_c = np.append(beta, c)
    fc, _, _ = accumulate(theta_c, cells)
    theta_d = np.append(beta, d)
    fd, _, _ = accumulate(theta_d, cells)

    while (b - a) > tol:
        if fc < fd:
            b = d; d = c; fd = fc
            c = b - phi * (b - a)
            theta_c = np.append(beta, c)
            fc, _, _ = accumulate(theta_c, cells)
        else:
            a = c; c = d; fc = fd
            d = a + phi * (b - a)
            theta_d = np.append(beta, d)
            fd, _, _ = accumulate(theta_d, cells)

    lam_opt = (a + b) / 2.0
    theta_opt = np.append(beta, lam_opt)
    negll_opt, _, _ = accumulate(theta_opt, cells)
    return lam_opt, negll_opt


# =========================================================================
# Coordinate descent: alternate β|λ and λ|β (matches Julia)
# =========================================================================

def coordinate_descent(cells, max_outer=20, outer_tol=1.0):

    print("  Coordinate descent: beta|lambda then lambda|beta")

    beta = np.zeros(K)
    lam = 1.0

    # Outer 0: optimize β at λ=1
    print(f"    Outer 0: optimizing beta at lambda = {lam:.4f}")
    beta, negll = optimize_beta(cells, beta_init=beta, lam_fixed=lam)
    print(f"    Outer 0: negLL = {negll:.2f}  lambda = {lam:.4f}  beta_1 = {beta[0]:.6f}")

    for outer in range(1, max_outer + 1):
        old_negll = negll

        # Optimize λ with β fixed
        lam, negll_lam = optimize_lambda(cells, beta)
        print(f"    Outer {outer}: lambda search -> lambda = {lam:.4f}  negLL = {negll_lam:.2f}")

        # Optimize β with λ fixed
        beta, negll = optimize_beta(cells, beta_init=beta, lam_fixed=lam)
        print(f"    Outer {outer}: beta optim  -> negLL = {negll:.2f}  beta_1 = {beta[0]:.6f}")

        if abs(negll - old_negll) < outer_tol:
            print(f"  Coordinate descent converged at outer iter {outer}")
            break

    print(f"  Final: negLL = {negll:.2f}  lambda = {lam:.4f}  beta_1 = {beta[0]:.6f}")
    return np.append(beta, lam), negll


# =========================================================================
# Main
# =========================================================================

def main():
    cell_dir = "data/output/choice_cells"
    out_dir = "results"

    print("=== Structural demand estimation (Python) ===")
    print("  V_0 = beta'X_0 (NOT 0)")
    print("  Coordinate descent: beta|lambda then lambda|beta")
    print(f"  Covariates: {K}")
    print()

    # Load cells
    cells, total_hh = load_cells(cell_dir)
    normalize_weights(cells)

    t0 = time.time()

    # Coordinate descent
    theta_opt, negll = coordinate_descent(cells)

    elapsed = time.time() - t0
    print(f"\n  Done in {elapsed / 60:.1f} min")
    print(f"  negLL = {negll:.2f}  lambda = {theta_opt[K]:.4f}")

    # Save coefficients
    terms = COVARS + ["lambda"]
    coef_path = os.path.join(out_dir, "choice_coefficients_structural.csv")
    with open(coef_path, "w", newline="") as f:
        writer = csv.writer(f)
        writer.writerow(["term", "estimate"])
        for t, e in zip(terms, theta_opt):
            writer.writerow([t, e])

    print("\n  Coefficients:")
    for t, e in zip(terms, theta_opt):
        print(f"    {t:25s} = {e:12.6f}")

    beta_p = theta_opt[COVARS.index("premium")]
    beta_c = theta_opt[COVARS.index("commission_broker")]
    if abs(beta_p) > 1e-10:
        print(f"\n  beta_commission / |beta_premium| = {beta_c / abs(beta_p):.4f}")

    print(f"  -> {coef_path}")
    print("\nDone.")


if __name__ == "__main__":
    main()

File: code/analysis/structural/estimate_demand.R
# Meta --------------------------------------------------------------------

## Author:        Ian McCarthy
## Date Created:  2026-03-28
## Description:   Nested logit demand estimation in R.
##                Faithful port of Julia estimate_demand_v3.jl.
##                Two-stage: (1) BFGS-BHHH on β with λ=1 fixed,
##                           (2) L-BFGS-B joint optimization.
##                Streams cells from .rds files — one cell in memory at a time.
##                Vectorized via rowsum() and matrix ops.
##
##                Key design:
##                  V_0 = β'X_0 (NOT normalized to 0)
##                  Weights normalized to mean 1 globally

# =========================================================================
# Covariates (shared between Phase 1 cell building and estimation)
# =========================================================================

STRUCTURAL_COVARS <- c(
  "premium", "penalty_own", "premium_sq",
  "silver", "bronze", "hh_size_prem",
  "any_0to17_prem", "FPL_250to400_prem", "FPL_400plus_prem",
  "any_black_prem", "any_hispanic_prem", "hmo", "hsa",
  "Anthem", "Blue_Shield", "Kaiser", "Health_Net",
  "Anthem_silver", "BS_silver", "Kaiser_silver", "HN_silver",
  "Anthem_bronze", "BS_bronze", "Kaiser_bronze", "HN_bronze",
  "assisted_silver", "assisted_bronze", "assisted_gold", "assisted_plat",
  "commission_broker", "v_hat_commission"
)


# =========================================================================
# Convert a choice data.frame into compact cell structure for estimation
# =========================================================================

build_cell_struct <- function(df, covars = STRUCTURAL_COVARS) {

  df <- as.data.table(df)
  setorder(df, household_number, plan_name)

  K <- length(covars)
  n <- nrow(df)

  X <- matrix(0, nrow = n, ncol = K)
  for (k in seq_along(covars)) {
    v <- covars[k]
    if (v %in% names(df)) {
      col <- df[[v]]
      col[is.na(col)] <- 0
      X[, k] <- as.numeric(col)
    }
  }

  hh_ids <- df$household_number
  plan_names <- df$plan_name
  choices <- df$choice
  weights <- df$ipweight

  breaks <- c(0L, which(diff(hh_ids) != 0), n)
  n_groups <- length(breaks) - 1L

  unins_row <- integer(n_groups)
  chosen_row <- integer(n_groups)
  chose_insured <- logical(n_groups)
  wt <- numeric(n_groups)
  valid <- logical(n_groups)
  ins_rows_list <- vector("list", n_groups)

  for (g in seq_len(n_groups)) {
    rows <- (breaks[g] + 1L):breaks[g + 1L]
    is_unins <- plan_names[rows] == "Uninsured"
    is_ins <- !is_unins
    is_chosen <- choices[rows] == 1L

    if (!any(is_ins) || !any(is_unins) || !any(is_chosen)) {
      valid[g] <- FALSE
      next
    }

    ins_rows_list[[g]] <- rows[is_ins]
    unins_row[g] <- rows[is_unins][1]
    chosen_row[g] <- rows[is_chosen][1]
    chose_insured[g] <- plan_names[chosen_row[g]] != "Uninsured"
    wt[g] <- weights[rows[is_ins][1]]
    valid[g] <- TRUE
  }

  keep <- which(valid)
  if (length(keep) == 0) return(NULL)

  n_hh <- length(keep)
  ins_rows_list <- ins_rows_list[keep]
  ins_rows_all <- unlist(ins_rows_list)
  group_id <- rep(seq_len(n_hh), lengths(ins_rows_list))

  list(
    X = X,
    ins_rows = ins_rows_all,
    group_id = group_id,
    unins_row = unins_row[keep],
    chosen_row = chosen_row[keep],
    chose_insured = chose_insured[keep],
    weight = wt[keep],
    n_hh = n_hh
  )
}


# =========================================================================
# NLL + gradient for one cell (fully vectorized)
#   compute_gradi = TRUE returns per-HH gradient matrix (for BHHH init)
# =========================================================================

cell_negll_grad <- function(beta, lambda, cell, compute_gradi = FALSE) {

  X <- cell$X
  K <- length(beta)
  n_hh <- cell$n_hh
  ins_rows <- cell$ins_rows
  gid <- cell$group_id
  w <- cell$weight
  ci <- as.numeric(cell$chose_insured)

  V_all <- as.numeric(X %*% beta)

  # --- Insured nest ---
  V_ins <- V_all[ins_rows]
  V_scaled <- pmin(pmax(V_ins / lambda, -500), 500)

  max_vs_per_hh <- as.numeric(tapply(V_scaled, gid, max))
  max_vs <- max_vs_per_hh[gid]

  exp_vs <- exp(V_scaled - max_vs)
  D_per_hh <- pmax(as.numeric(rowsum(cbind(exp_vs), gid)), 1e-300)
  D <- D_per_hh[gid]

  I_val <- max_vs_per_hh + log(D_per_hh)
  s_jg <- exp_vs / D
  s_jg[!is.finite(s_jg)] <- 0

  X_ins <- X[ins_rows, , drop = FALSE]
  x_bar <- rowsum(X_ins * s_jg, gid)
  V_bar <- as.numeric(rowsum(cbind(s_jg * V_ins), gid))

  # --- Outside option ---
  V_0 <- V_all[cell$unins_row]
  X_0 <- X[cell$unins_row, , drop = FALSE]

  # --- Nest probability ---
  lI <- lambda * I_val
  mx_d <- pmax(lI, V_0)
  log_denom <- mx_d + log(exp(pmin(lI - mx_d, 500)) + exp(pmin(V_0 - mx_d, 500)))
  P_ins <- pmin(pmax(exp(lI - log_denom), 1e-15), 1 - 1e-15)

  # --- Log-likelihood ---
  V_ch <- V_all[cell$chosen_row]
  ll_ins <- V_ch / lambda + (lambda - 1) * I_val - log_denom
  ll_out <- V_0 - log_denom
  ll_h <- ci * ll_ins + (1 - ci) * ll_out
  negll <- -sum(w * ll_h)

  # --- Gradient w.r.t. β (per-HH matrix) ---
  X_ch <- X[cell$chosen_row, , drop = FALSE]
  diff_xbar_x0 <- x_bar - X_0
  g_beta_mat <- (ci / lambda) * (X_ch - x_bar) + (ci - P_ins) * diff_xbar_x0
  grad_beta <- -colSums(w * g_beta_mat)

  # --- Gradient w.r.t. λ ---
  I_minus_Vbar_lam <- I_val - V_bar / lambda
  g_lam_vec <- ci * (-V_ch / lambda^2 + I_val - (lambda - 1) * V_bar / lambda^2) -
    P_ins * I_minus_Vbar_lam
  grad_lambda <- -sum(w * g_lam_vec)

  # Guard against NaN
  if (!is.finite(negll)) negll <- 1e20
  grad <- c(grad_beta, grad_lambda)
  grad[!is.finite(grad)] <- 0

  # Per-HH gradient matrix (for BHHH)
  gradi <- NULL
  if (compute_gradi) {
    gradi <- cbind(-w * g_beta_mat, -w * g_lam_vec)
  }

  list(negll = negll, grad = grad, gradi = gradi)
}


# =========================================================================
# Streaming accumulate: NLL + gradient (one cell at a time)
# =========================================================================

accumulate_streaming <- function(theta, rds_files, weight_scale) {
  K <- length(theta) - 1L
  beta <- theta[1:K]
  lambda <- theta[K + 1L]

  total_negll <- 0
  total_grad <- numeric(K + 1L)

  for (i in seq_along(rds_files)) {
    cell <- readRDS(rds_files[i])
    cell$weight <- cell$weight / weight_scale
    res <- cell_negll_grad(beta, lambda, cell)
    total_negll <- total_negll + res$negll
    total_grad <- total_grad + res$grad
    if (i %% 20 == 0) gc(verbose = FALSE)
  }

  list(negll = total_negll, grad = total_grad)
}


# =========================================================================
# Streaming accumulate with BHHH: NLL + gradient + BHHH matrix
# =========================================================================

accumulate_streaming_bhhh <- function(theta, rds_files, weight_scale) {
  K <- length(theta) - 1L
  beta <- theta[1:K]
  lambda <- theta[K + 1L]

  total_negll <- 0
  total_grad <- numeric(K + 1L)
  bhhh <- matrix(0, K, K)

  for (i in seq_along(rds_files)) {
    cell <- readRDS(rds_files[i])
    cell$weight <- cell$weight / weight_scale
    res <- cell_negll_grad(beta, lambda, cell, compute_gradi = TRUE)
    total_negll <- total_negll + res$negll
    total_grad <- total_grad + res$grad
    gi_beta <- res$gradi[, 1:K, drop = FALSE]
    bhhh <- bhhh + crossprod(gi_beta)
    if (i %% 20 == 0) gc(verbose = FALSE)
  }

  list(negll = total_negll, grad = total_grad, bhhh = bhhh)
}


# =========================================================================
# Stage 1: BFGS-BHHH on β with λ fixed (matches Julia optimize_beta)
# =========================================================================

optimize_beta_streaming <- function(rds_files, weight_scale, K,
                                     beta_init = NULL, lambda_fixed = 1.0,
                                     max_iter = 2000, tol = 1e-6,
                                     stptol = 1e-10) {

  beta <- if (is.null(beta_init)) rep(0, K) else beta_init
  theta <- c(beta, lambda_fixed)

  # Initial evaluation with BHHH
  cat("  Computing BHHH initialization...\n")
  res <- accumulate_streaming_bhhh(theta, rds_files, weight_scale)
  negll <- res$negll
  grad_beta <- res$grad[1:K]

  Hm1 <- tryCatch(solve(res$bhhh), error = function(e) {
    warning("BHHH singular, using scaled identity")
    diag(1e-4, K)
  })

  cat(sprintf("  Init: negLL = %.2f  beta_1 = %.6f\n", negll, beta[1]))

  for (iter in seq_len(max_iter)) {
    old_negll <- negll
    old_grad <- grad_beta

    d <- -as.numeric(Hm1 %*% grad_beta)

    # Halving line search
    step <- 2.0
    repeat {
      step <- step / 2
      if (step < stptol) break
      theta_try <- c(beta + step * d, lambda_fixed)
      res_try <- accumulate_streaming(theta_try, rds_files, weight_scale)
      if (res_try$negll <= old_negll) break
    }

    if (step < stptol) {
      cat(sprintf("    beta iter %3d: step too small\n", iter))
      break
    }

    beta <- beta + step * d
    theta <- c(beta, lambda_fixed)
    res <- accumulate_streaming(theta, rds_files, weight_scale)
    negll <- res$negll
    grad_beta <- res$grad[1:K]

    # BFGS inverse Hessian update
    incr <- step * d
    y <- grad_beta - old_grad
    sy <- sum(incr * y)
    if (abs(sy) > 1e-20) {
      yHy <- as.numeric(crossprod(y, Hm1 %*% y))
      Hm1 <- Hm1 + (sy + yHy) / sy^2 * tcrossprod(incr) -
        (Hm1 %*% tcrossprod(y, incr) + tcrossprod(incr, y) %*% Hm1) / sy
    }

    if (iter %% 10 == 0 || iter <= 3) {
      cat(sprintf("    beta iter %3d: negLL = %.2f  step = %.4f  beta_1 = %.6f\n",
                  iter, negll, step, beta[1]))
    }

    if (abs(negll - old_negll) < tol) {
      cat(sprintf("    beta converged at iter %d\n", iter))
      break
    }
  }

  list(beta = beta, negll = negll)
}


# =========================================================================
# Main estimation (two-stage, matching Julia main())
# =========================================================================

estimate_structural_demand <- function(cell_dir, out_dir = "results",
                                       start = NULL, max_iter = 2000) {

  K <- length(STRUCTURAL_COVARS)

  rds_files <- sort(list.files(cell_dir, pattern = "\\.rds$", full.names = TRUE))
  if (length(rds_files) == 0) stop("No .rds cell files found in ", cell_dir)

  cat("=== Structural demand estimation ===\n")
  cat("  V_0 = beta'X_0 (NOT 0)\n")
  cat("  Two-stage: BFGS-BHHH (beta|lambda=1) then L-BFGS-B (joint)\n")
  cat("  Streaming cells from .rds (one at a time)\n")
  cat("  Covariates:", K, ", Cells:", length(rds_files), "\n\n")

  # Compute global weight mean
  cat("  Computing weight normalization...\n")
  total_w <- 0
  total_hh <- 0L
  for (f in rds_files) {
    cell <- readRDS(f)
    total_w <- total_w + sum(cell$weight)
    total_hh <- total_hh + cell$n_hh
  }
  rm(cell); gc(verbose = FALSE)
  weight_scale <- total_w / total_hh
  cat("  HH:", total_hh, " Weight scale:", round(weight_scale, 4), "\n")

  t0 <- proc.time()

  if (!is.null(start)) {
    # Warm start: skip Stage 1
    theta_start <- start
    cat("\n  Warm start from provided coefficients — skipping Stage 1\n")
  } else {
    # Stage 1: BFGS-BHHH on β with λ=1 fixed
    cat("\n  Stage 1: MNL starting values (beta|lambda=1)...\n")
    stage1 <- optimize_beta_streaming(rds_files, weight_scale, K)
    theta_start <- c(stage1$beta, 1.0)
    cat(sprintf("  Stage 1 done: negLL = %.2f\n", stage1$negll))
    rm(stage1); gc(verbose = FALSE)
  }

  # Stage 2: L-BFGS-B joint optimization
  cat("\n  Stage 2: L-BFGS-B joint optimization...\n")

  cache <- new.env(parent = emptyenv())
  cache$theta <- NULL
  cache$result <- NULL
  cache$n_eval <- 0L

  eval_cached <- function(theta) {
    theta[K + 1L] <- max(min(theta[K + 1L], 5.0), 0.001)
    if (!identical(theta, cache$theta)) {
      cache$theta <- theta
      cache$result <- accumulate_streaming(theta, rds_files, weight_scale)
      cache$n_eval <- cache$n_eval + 1L
      if (cache$n_eval %% 50 == 0) {
        cat(sprintf("  eval %d: negLL = %.2f  lambda = %.4f\n",
                    cache$n_eval, cache$result$negll, theta[K + 1L]))
      }
    }
    cache$result
  }
  fn <- function(theta) eval_cached(theta)$negll
  gr <- function(theta) eval_cached(theta)$grad

  result <- optim(
    par = theta_start,
    fn = fn,
    gr = gr,
    method = "L-BFGS-B",
    lower = c(rep(-Inf, K), 0.001),
    upper = c(rep(Inf, K), 5.0),
    control = list(maxit = max_iter, factr = 1e-10, pgtol = 1e-8,
                   trace = 0)
  )

  elapsed <- (proc.time() - t0)[3]
  cat(sprintf("\n  Done in %.1f min\n", elapsed / 60))
  cat(sprintf("  negLL = %.2f  convergence = %d  lambda = %.4f\n",
              result$value, result$convergence, result$par[K + 1L]))

  # Save coefficients
  coefs <- data.frame(
    term = c(STRUCTURAL_COVARS, "lambda"),
    estimate = result$par
  )
  out_path <- file.path(out_dir, "choice_coefficients_structural.csv")
  write.csv(coefs, out_path, row.names = FALSE)

  cat("\n  Coefficients:\n")
  for (i in seq_len(nrow(coefs))) {
    cat(sprintf("    %-25s = %12.6f\n", coefs$term[i], coefs$estimate[i]))
  }

  beta_p <- coefs$estimate[coefs$term == "premium"]
  beta_c <- coefs$estimate[coefs$term == "commission_broker"]
  if (abs(beta_p) > 1e-10) {
    cat(sprintf("\n  beta_commission / |beta_premium| = %.4f\n", beta_c / abs(beta_p)))
  }

  cat("  ->", out_path, "\n")
  coefs
}
